library/ssl_tls12_server.c

```diff
@@ -2804,6 +2804,14 @@ static int ssl_write_certificate_request( mbedtls_ssl_context *ssl )
          *       `mbedtls_ssl_conf_ca_cb()`, then the
          *       CertificateRequest is currently left empty. */
 
+#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
```

Comment:
:rocket:
I'll have to trace back to try to remember why I did this.  It may have been related to my first implementation attempt.  I have changed it to check `#if defined(MBEDTLS_KEY_EXCHANGE_CERT_REQ_ALLOWED_ENABLED)`, which wraps access to the member in `mbedtls_ssl_handshake_params`

Reply:
If I recall correctly, the reason I added the check `#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)` was a possible small optimization: if you configure dn_hint with `mbedtls_ssl_config`, why might you need to change it per-handshake if the handshake did not include SNI?

Reply:
Answer: `mbedtls_ssl_set_hs_*` prototypes are all protected by `#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)` in `include/mbedtls/ssl.h`.  If this should not be the case, then I also need to move the prototype in for `mbedtls_ssl_set_hs_dn_hints()` in `include/mbedtls/ssl.h`
:rocket:

Reply:
I am hoping this PR will be included in mbedtls 3.2.0, so I think filing a separate issue is the place to consider expanding access.  Personally, I think that the dependency on SNI make sense.  I think that without SNI used to make a different choice, you might as well configure these in `set_conf` rather than in `set_hs`.

Would you please create a new issue to describe what you would like to see changed regarding `set_hs` and `MBEDTLS_SSL_SERVER_NAME_INDICATION`?  (Subscribe me and I might try to mock it up in a new PR.)

Reply:
@mpg - gentle reminder to do this, if not already done.